### PR TITLE
Split cache step into two, always save cache, fix hash files

### PR DESCRIPTION
**Pull request by EnricoMi**

### CODE REVIEW

File: composite/action.yml [hunk 1]
```diff
@@ -238,6 +239,14 @@ runs:
         LOG_LEVEL: ${{ inputs.log_level }}
       shell: bash
 
+    - name: Save PIP packages cache
+      uses: actions/cache/save@v3
+      if: always()
```
Comment (TWiStErRob) [inline composite/action.yml:244]: with this, you can essentially remove the date trick that accounts for https://github.com/EnricoMi/publish-unit-test-result-action/pull/418#discussion_r1135436316, right?

File: composite/action.yml [hunk 2]
```diff
@@ -175,8 +175,9 @@ runs:
         esac
       shell: bash
 
-    - name: Cache PIP Packages
-      uses: actions/cache@v3
+    - name: Restore PIP packages cache
+      id: cache
+      uses: actions/cache/restore@v3
```
Comment (TWiStErRob) [inline composite/action.yml:180]: can I use this branch directly?

Comment (TWiStErRob) [reply composite/action.yml:180]: Let's see https://github.com/TWiStErRob/net.twisterrob.libraries/pull/38/commits/b66a8c6c695b88f4e359c275b9298d36bdb7a4cd

File: composite/action.yml (hunk 1 above)
Comment (TWiStErRob) [inline composite/action.yml:244]: `if: success() || failure()`

`always()` includes `cancelled()` which might result in a partial broken cache

https://docs.github.com/en/actions/learn-github-actions/expressions#status-check-functions

File: composite/action.yml (hunk 2 above)
Comment (EnricoMi) [reply composite/action.yml:183]:
```
uses: EnricoMi/publish-unit-test-result-action/composite@branch-composite-cache-pip-packages-2
```

**Comment by TWiStErRob** [comment]
Please let me know if I should re-run on latest of this PR. Also please see the comments in this PR, they might be hiding between the test results.

**Comment by EnricoMi** [comment]
@TWiStErRob go for it, please test **this branch**!

**Comment by TWiStErRob** [comment]
Queued https://github.com/TWiStErRob/net.twisterrob.libraries/actions/runs/4416412575/jobs/7769498594

**Comment by TWiStErRob** [comment]
😍  image

**Comment by TWiStErRob** [comment]
I reran it just to try and it took more than a minute second time:
https://github.com/TWiStErRob/net.twisterrob.libraries/actions/runs/4416412575/attempts/4

Both times it restored caches.
I cleared caches and running it again:
https://github.com/TWiStErRob/net.twisterrob.libraries/actions/runs/4416412575
and will run again when it finishes to see the cache-reuse.

**Comment by EnricoMi** [comment]
Can you rerun the `🧪 Instrumentation Tests on 21` job, please? There is still an issue with the cache key.

**Comment by TWiStErRob** [comment]
Cancelled and running again with latest: https://github.com/TWiStErRob/net.twisterrob.libraries/actions/runs/4416412575/attempts/6
invited you, feel free to edit the branch and run the CI directly on https://github.com/TWiStErRob/net.twisterrob.libraries/pull/38

**Comment by EnricoMi** [comment]
> invited you

Thanks!

**Comment by EnricoMi** [comment]
Run with cache miss:
https://github.com/TWiStErRob/net.twisterrob.libraries/actions/runs/4429767427/jobs/7790625895#step:6:92

Run with cache hit:
https://github.com/TWiStErRob/net.twisterrob.libraries/actions/runs/4429767427/jobs/7791493539#step:6:96

File: composite/action.yml [hunk 3]
```diff
@@ -175,12 +175,13 @@ runs:
...
-    - name: Cache PIP Packages
-      uses: actions/cache@v3
+    - name: Restore PIP packages cache
+      uses: actions/cache/restore@v3
+      id: cache
       continue-on-error: true
       with:
         path: ${{ steps.os.outputs.pip-cache }}
-        key: enricomi-publish-action-${{ runner.os }}-${{ runner.arch }}-pip-${{ steps.python.outputs.version }}-${{ hashFiles('**/requirements.txt', 'composite/action.yml') }}
+        key: enricomi-publish-action-${{ runner.os }}-${{ runner.arch }}-pip-${{ steps.python.outputs.version }}-f3f2c295046c91ed612b4efb6c9fb352
```
Comment (TWiStErRob) [inline composite/action.yml:184]: I saw you were fighting with this, in the end you'll update it manually for each release?

**Comment by TWiStErRob** [comment]
Amazing! Congrats, it looked like a very hard labor. It's funny how 15MB can cut 3.5min to 15 seconds

File: composite/action.yml (hunk 3 above)
Comment (TWiStErRob) [reply composite/action.yml:184]: it's in `glob`: https://github.com/actions/toolkit/blob/main/packages/glob/src/glob.ts#L29

Comment (EnricoMi) [reply composite/action.yml:184]: > you'll update it manually for each release

yes, but only if `requirements.txt` changes

Comment (EnricoMi) [reply composite/action.yml:184]: > and call `hashFiles` directly from JS?

maybe, but I'd prefer not to add more complexity

File: composite/action.yml [hunk 4]
```diff
@@ -175,12 +175,13 @@ runs:
...
 
     - name: Install Python dependencies
```
[the remark this answers is not in the document]
Comment (EnricoMi) [reply composite/action.yml:186]: Don't get it. With `key: ${{ steps.cache.outputs.cache-primary-key }}`, the same key is used when saving the cache. Why should the python version change?

**Pull request merged**